Docker image for Overpass

wiktorn opened the pull request:
Based on NGINX image. Uses supervisord to run:
* nginx
* fastcgi wrapper
* Overpass dispatcher
* fetch_osc.sh
* apply_osc_to_db.sh

On first run initializes itself with the data. Details provided in README_Docker.md.

This depends on changes in: #373. `apply_osc_to_db.sh` may start before `dispatcher` is ready. Thanks to changes in #373 it will die and start again, when `dispatcher` is fully working.

Changes in src/bin/fetch_osc.sh skips the download of files already available locally on restart of the container.

drolbr committed:
Fixed bad resource management in out:popup

drolbr committed:
Fixed further bug in out:popup

wiktorn committed:
Docker image for Overpass

Based on NGINX image. Uses supervisord to run:
* nginx
* fastcgi wrapper
* Overpass dispatcher
* fetch_osc.sh
* apply_osc_to_db.sh

On first run initializes itself with the data. Details provided in README_Docker.md.

wiktorn committed:
Add quotations around variable name

mmd-osm commented:
Needs cross checking with what has already been done in https://github.com/drolbr/Overpass-API/issues/318 and the ~ 5 other Docker images proposed so far.

wiktorn commented:
I was using some of the Dockerfiles as inspiration. The only part that I'm aware,  that's missing is:
- area creation
- attic support

If for areas support I need two extra processes:
* `$EXEC_DIR/bin/dispatcher --areas --db-dir=$DB_DIR`
* `EXEC_DIR/bin/rules_loop.sh $DB_DIR`

And populate `$DB_DIR` with src/rules on initialization, then it looks like an easy way forward.

mmd-osm commented:
Compression settings could also be interesting beyond area and attic support.

wiktorn committed:
Merge remote-tracking branch 'upstream/master' into docker_image

wiktorn committed:
Add support for attic, areas and compression options.

Add liblz4 to image so Overpass will support additional compression algorithm.
Add two processes:
* dispatcher_areas - for calculating areas
* areas_rules - rules loop

Added src/rules to databases directory

Added OVERPASS_COMPRESSION environment variable

Added `attic` value to OVERPASS_META variable

wiktorn committed:
Merge branch 'docker_image' of github.com:wiktorn/Overpass-API into docker_image

wiktorn committed:
Add --enable-lz4 to configure script

wiktorn commented:
I've just added area creation, attic support.

I've tried added compression but though i added --enable-lz4 and it looks like it's properly identified, bin/update_database doesn't report --compression=lz4.
configure log:
```
checking lz4.h usability... yes
checking lz4.h presence... yes
checking for lz4.h... yes
checking for LZ4_compress in -llz4... yes
```

```
$ cat /app/src/config.h | grep LZ4
#define HAVE_LZ4 1

$ /app/bin/update_database -h
Unkown argument: -h
Usage: /app/bin/update_database [--db-dir=DIR] [--version=VER] [--meta|--keep-attic] [--flush_size=FLUSH_SIZE] [--compression_method=(no|gz)]
```

mmd-osm commented:
> bin/update_database doesn't report --compression=lz4.

I think that's a bug the coding (https://github.com/drolbr/Overpass-API/blob/a3ab379733cbf89835a1722204d7a5fe459209aa/src/overpass_api/osm-backend/update_database.cc#L114), the help/usage text doesn't correctly reflect the available compression options.

wiktorn committed:
Copy areas rules to database folder

wiktorn committed:
Add rules load to supervisord

mmd-osm commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+            META="--meta=$OVERPASS_META"
+        fi
+        mkdir -p /db/db \
+        && /app/bin/download_clone.sh --db-dir=/db/db --source=http://dev.overpass-api.de/api_drolbr/ $META "--compresion_method=$OVERPASS_COMPRESSION" \
```
download_clone.sh just downloads existing files with a predefined compression setting, i.e.  there's not much point to provide a compression method parameter, as it will not have any impact.

mmd-osm commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,36 @@
+
+    if [ "$OVERPASS_MODE" = "init" ]; then
+        lftp -c "get -c \"$OVERPASS_PLANET_URL\" -o /db/planet; exit" \
+        && /app/bin/init_osm3s.sh /db/planet /db/db /app "--meta=$OVERPASS_META" \
```
compression setting is better suited here.

wiktorn committed:
Fix passing compression algorithm.

`OVERPASS_COMPRESSION` makes sense only for init_osm3s not for download_clone.

Reported by: @mmd-osm

wiktorn commented:
God catch, thanks.

mmd-osm commented on docker-entrypoint.sh:
```diff
@@ -23,7 +23,7 @@ if [ ! -d /db/db ] ; then
     if [ "$OVERPASS_MODE" = "init" ]; then
         lftp -c "get -c \"$OVERPASS_PLANET_URL\" -o /db/planet; exit" \
-        && /app/bin/init_osm3s.sh /db/planet /db/db /app "--meta=$OVERPASS_META" \
+        && /app/bin/init_osm3s.sh /db/planet /db/db /app "--meta=$OVERPASS_META" "--compresion_method=$OVERPASS_COMPRESSION" \
```
... and the parameter should be `--compression-method=` with a dash (not an underscore, the usage text is wrong).

wiktorn replied:
Ok, I'll create a seperate pull request for usage text

wiktorn replied:
#381 it is

wiktorn committed:
Fix spelling of parameter

wiktorn committed:
Add gecos parameter to adduser

wiktorn committed:
Merge remote-tracking branch 'upstream/master' into docker_image

wiktorn committed:
Add documentation about OVERPASS_RULES_LOAD

wiktorn committed:
Fix typo

gy-mate commented:
@wiktorn @mmd-osm @drolbr Why was this PR closed? Were there any unresolved issues with it?

I'd love to see Docker support!
